airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcDestinationHandler.java

```diff
@@ -184,4 +210,55 @@ public static Optional<TableDefinition> findExistingTable(final JdbcDatabase jdb
     return Optional.of(new TableDefinition(retrievedColumnDefns));
   }
 
+  public static boolean fromIsNullableIsoString(final String isNullable) {
+    return "YES".equalsIgnoreCase(isNullable);
+  }
+
+  private boolean isAirbyteRawIdColumnMatch(final TableDefinition existingTable) {
+    return existingTable.columns().containsKey(COLUMN_NAME_AB_RAW_ID) &&
+        toJdbcTypeName(AirbyteProtocolType.STRING).equals(existingTable.columns().get(COLUMN_NAME_AB_RAW_ID).type());
+  }
+
+  private boolean isAirbyteExtractedAtColumnMatch(final TableDefinition existingTable) {
+    return existingTable.columns().containsKey(COLUMN_NAME_AB_EXTRACTED_AT) &&
+        toJdbcTypeName(AirbyteProtocolType.TIMESTAMP_WITH_TIMEZONE).equals(existingTable.columns().get(COLUMN_NAME_AB_EXTRACTED_AT).type());
+  }
+
+  private boolean isAirbyteMetaColumnMatch(final TableDefinition existingTable) {
+    return existingTable.columns().containsKey(COLUMN_NAME_AB_META) &&
+        toJdbcTypeName(new Struct(new LinkedHashMap<>())).equals(existingTable.columns().get(COLUMN_NAME_AB_META).type());
+  }
+
+  protected boolean existingSchemaMatchesStreamConfig(final StreamConfig stream, final TableDefinition existingTable) {
+    // Check that the columns match, with special handling for the metadata columns.
+    if (!isAirbyteRawIdColumnMatch(existingTable) ||
+        !isAirbyteExtractedAtColumnMatch(existingTable) ||
+        !isAirbyteMetaColumnMatch(existingTable)) {
+      // Missing AB meta columns from final table, we need them to do proper T+D so trigger soft-reset
+      return false;
+    }
+    final LinkedHashMap<String, String> intendedColumns = stream.columns().entrySet().stream()
+        .collect(LinkedHashMap::new,
+            (map, column) -> map.put(column.getKey().name(), toJdbcTypeName(column.getValue())),
+            LinkedHashMap::putAll);
+
+    // Filter out Meta columns since they don't exist in stream config.
+    final LinkedHashMap<String, String> actualColumns = existingTable.columns().entrySet().stream()
+        .filter(column -> V2_FINAL_TABLE_METADATA_COLUMNS.stream()
+            .noneMatch(airbyteColumnName -> airbyteColumnName.equals(column.getKey())))
+        .collect(LinkedHashMap::new,
+            (map, column) -> map.put(column.getKey(), column.getValue().type()),
+            LinkedHashMap::putAll);
+
+    return actualColumns.equals(intendedColumns);
+  }
+
+  /**
+   * Convert to the TYPE_NAME retrieved from {@link java.sql.DatabaseMetaData#getColumns}
+   *
+   * @param airbyteType
+   * @return
+   */
+  protected abstract String toJdbcTypeName(final AirbyteType airbyteType);
```

Comment:
feels kind of sad that we need this when JdbcSqlGenerator already has `DataType<?> toDialectType(final AirbyteProtocolType airbyteProtocolType)` :cry: but I don't see a great way to avoid it b/c of jooq vs jdbc differences

though... I kind of wonder if it's easier to just pass a `Function<AIrbyteType, String>` into the JdbcDestinationHandler constructor, rather than defining this as an abstract method. Then we'd just need to override PostgresDestination#getDestinationHandler, without needing to define an entire new class

Reply:
Yeah didn't spend enough time to pull all those out to a Functional model for late binding the call.